File: app/src/main/java/com/shixun/android/leaving_detection/Detection/WifiScanRunnable.java
package com.shixun.android.leaving_detection.Detection;

import android.net.wifi.ScanResult;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;

import org.litepal.crud.DataSupport;

import java.util.List;

/**
 * Created by shixunliu on 22/4/17.
 */

public class WifiScanRunnable implements Runnable {

    private WifiManager mWifiManager = null;
    private int dataSize = 10;
    private String[] strArray;

    public WifiScanRunnable(WifiManager wifiManager, String[] strArray) {
        this.mWifiManager = wifiManager;
        this.strArray = strArray;
    }

    @Override
    public void run() {
        scanAndSaveWifiData();
        if (DataSupport.count(WifiData.class) >= dataSize) {
            strArray[1] = (" 30:" + getMean("homeWifiLevel", WifiData.class)
                    + " 31:" + getMean("meanOfAllWifiLevel", WifiData.class))
                    + " 32:" + getStd("meanOfHomeWifi")
                    + " 33:" + getSumVar("meanOfHomeWifi");
                    //+ " 22:" + getSumVar("meanOfAllWifiLevel");
            //删除一组数据
            int id = DataSupport.findFirst(WifiData.class).getId();
            DataSupport.delete(WifiData.class, id);
        }
    }

    private void scanAndSaveWifiData() {

        WifiInfo wifiInfo = null;
        double isHomeWifi = 0;
        double allWifiLevel = 0;
        double meanOfAllWifi = 0;
        double homeWifiLevel = 0;

        mWifiManager.startScan();
        wifiInfo = mWifiManager.getConnectionInfo();

        homeWifiLevel = wifiInfo.getRssi();
        if (Math.abs(homeWifiLevel) >= 95) {
            homeWifiLevel = -95;
            isHomeWifi = 0.0;
        } else {
            isHomeWifi = 1.0;
        }

        List<ScanResult> scanResults = mWifiManager.getScanResults();
        for (ScanResult scanResult : scanResults) {
            allWifiLevel = allWifiLevel + scanResult.level;
        }
        meanOfAllWifi = allWifiLevel / scanResults.size();

        WifiData sample = new WifiData();
        sample.setHomeWifiLevel(homeWifiLevel);
        sample.setMeanOfAllWifiLevel(meanOfAllWifi);
        sample.setIsHomeWifi(isHomeWifi);
        sample.setStdOfAllWifiLevel(meanOfAllWifi);
        if(DataSupport.count(WifiData.class) == 0){
            sample.setMeanOfHomeWifi(homeWifiLevel);
        } else {
            sample.setMeanOfHomeWifi(getMean("homeWifiLevel", WifiData.class));
        }
        sample.saveThrows();
    }

    private double getSumVar(String col) {
        List<WifiData> lastSet;
        lastSet = DataSupport.select(col).order("id desc").limit(5).find(WifiData.class);

        double max = lastSet.get(0).getMeanOfHomeWifi();
        double min = lastSet.get(4).getMeanOfHomeWifi();

        return max - min;
    }

    private double getMean(String col, Class className) {
        double mean = DataSupport.average(className, col);
        return mean;
    }

    private double getStd(String col) {
        List<WifiData> temp = DataSupport.select(col).find(WifiData.class);
        double mean = DataSupport.average(WifiData.class, col);
        double result = 0;
        for (int i = 0; i < temp.size() - 1; i++) {
            result += Math.pow(temp.get(i).getMeanOfHomeWifi() - mean, 2);
        }
        return Math.sqrt(result / (temp.size() - 2));
    }
}

File: app/src/main/java/com/shixun/android/leaving_detection/DataCollection/WifiScanRunnable.kt
package com.shixun.android.leaving_detection.DataCollection

import android.net.wifi.WifiManager
import org.litepal.crud.DataSupport

/**
 * Created by shixunliu on 22/4/17.
 */

class WifiScanRunnable(val wifiManager: WifiManager, private val strArray: Array<String>, private val doubleArray: DoubleArray) : Runnable {

    private val dataSize = 20

    override fun run() {
        scanAndSaveWifiData()
        //val currentValue = scanAndSaveWifiData()
        if (DataSupport.count(WifiData::class.java) >= dataSize) {
            val currentValue = "%.3f".format(getMean("homeWifiLevel", WifiData::class.java))
            val std = "%.3f".format(getStd("meanOfHomeWifi"))
            val sumVar = "%.3f".format(getSumVar("meanOfHomeWifi"))
            doubleArray[1] = currentValue.toDouble()
            strArray[1] = " 30:$currentValue 31:$std 32:$sumVar"

            val lastId = DataSupport.findFirst<WifiData>(WifiData::class.java).id
            DataSupport.delete(WifiData::class.java, lastId.toLong())
        }
    }

    private fun scanAndSaveWifiData(): Double {

//        var allWifiLevel = 0.0
        wifiManager.startScan()
        val wifiInfo = wifiManager.connectionInfo
        var homeWifiLevel = wifiInfo!!.rssi.toDouble()
//        val scanResults = wifiManager.scanResults

//        scanResults.forEach {
//            allWifiLevel += it.level
//        }
//        val meanOfAllWifi = allWifiLevel / scanResults.size

        val sample = WifiData()
        sample.homeWifiLevel = homeWifiLevel
//        sample.meanOfAllWifiLevel = meanOfAllWifi
//        sample.stdOfAllWifiLevel = meanOfAllWifi
        if (DataSupport.count(WifiData::class.java) == 0) {
            sample.meanOfHomeWifi = homeWifiLevel
        } else {
            sample.meanOfHomeWifi = getMean("homeWifiLevel", WifiData::class.java)
        }
        sample.saveThrows()
        return homeWifiLevel
    }

    private fun getSumVar(col: String): Double {
        val lastSet: List<WifiData>
        //lastSet = DataSupport.select(col).order("id desc").limit(10).find(WifiData.class);
        lastSet = DataSupport.select(col).find<WifiData>(WifiData::class.java)
        val size = lastSet.size

        val max = lastSet[size - 1].meanOfHomeWifi
        val min = lastSet[0].meanOfHomeWifi

        return max - min
    }

    private fun getMean(col: String, className: Class<*>): Double {
        val mean = DataSupport.average(className, col)
        return mean
    }

    private fun getStd(col: String): Double {
        val temp = DataSupport.select(col).find<WifiData>(WifiData::class.java)
        val mean = DataSupport.average(WifiData::class.java!!, col)
        var result = 0.0
        for (i in 0..temp.size - 1 - 1) {
            result += Math.pow(temp[i].meanOfHomeWifi!! - mean, 2.0)
        }
        return Math.sqrt(result / (temp.size - 2))
    }
}
